### PR TITLE
SAAS-2231: Plugin to check access of AWS Support Center via IAM policy

### PR DESCRIPTION
**Pull request by UmairAjmal**

### CODE REVIEW

File: exports.js [hunk 1]
```diff
@@ -235,16 +237,17 @@ module.exports = {
         'usersPasswordAndKeys'          : require(__dirname + '/plugins/aws/iam/usersPasswordAndKeys.js'),
         'usersPasswordLastUsed'         : require(__dirname + '/plugins/aws/iam/usersPasswordLastUsed.js'),
         'canaryKeysUsed'                : require(__dirname + '/plugins/aws/iam/canaryKeysUsed.js'),
+        'iamPoliciesPresent'            : require(__dirname + '/plugins/aws/iam/iamPoliciesPresent.js'),
+        'iamDbAuthenticationEnabled'    : require(__dirname + '/plugins/aws/rds/iamDbAuthenticationEnabled.js'),
+        'iamSupportPolicy'                : require(__dirname + '/plugins/aws/iam/iamSupportPolicy.js'),
```
Comment (AkhtarAmir) [inline exports.js:242]:
```suggestion
        'iamSupportPolicy'              : require(__dirname + '/plugins/aws/iam/iamSupportPolicy.js'),
```

File: plugins/aws/iam/iamSupportPolicy.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    recommended_action: 'Ensure that an IAM role has permission to access support center.',
+    apis: ['IAM:listPolicies', 'IAM:listEntitiesForPolicy'],
+    settings: {},
+
+    run: function(cache, settings, callback) {
+
+        var results = [];
+        var source = {};
+
+        var region = helpers.defaultRegion(settings);
```
Comment (AkhtarAmir) [inline plugins/aws/iam/iamSupportPolicy.js:18]:
```suggestion
        var results = [];
        var source = {};
        var region = helpers.defaultRegion(settings);
```

File: plugins/aws/iam/iamSupportPolicy.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+        var policyArn = '';
+        listPolicies.data.forEach(policy => {
+            if (policy.PolicyName == 'AWSSupportAccess') {
+                if (policy.AttachmentCount > 0) {
+                    policyArn = policy.Arn;
+                } else {
+                    helpers.addResult(results, 2,
+                        'No role, user or group attached to the policy', 'global', policy.Arn);
+                }
```
Comment (AkhtarAmir) [inline plugins/aws/iam/iamSupportPolicy.js:40]: return callback(null, results, source)

File: plugins/aws/iam/iamSupportPolicy.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        }
+
+        var policyArn = '';
+        listPolicies.data.forEach(policy => {
+            if (policy.PolicyName == 'AWSSupportAccess') {
+                if (policy.AttachmentCount > 0) {
+                    policyArn = policy.Arn;
+                } else {
+                    helpers.addResult(results, 2,
+                        'No role, user or group attached to the policy', 'global', policy.Arn);
```
Comment (AkhtarAmir) [inline plugins/aws/iam/iamSupportPolicy.js:39]:
```suggestion
                        'No role, user or group attached to AWSSupportAccess policy', 'global', policy.Arn);
```

File: plugins/aws/iam/iamSupportPolicy.js [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        var results = [];
+        var source = {};
+
+        var region = helpers.defaultRegion(settings);
+
+        const listPolicies = helpers.addSource(cache, source,
+            ['iam', 'listPolicies', region]);
+        
+        const listEntitiesForPolicy = helpers.addSource(cache, source,
+            ['iam', 'listEntitiesForPolicy', region]);
```
Comment (AkhtarAmir) [inline plugins/aws/iam/iamSupportPolicy.js:22]:
```suggestion
```

File: plugins/aws/iam/iamSupportPolicy.js [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+                } else {
+                    helpers.addResult(results, 2,
+                        'No role, user or group attached to the policy', 'global', policy.Arn);
+                }
+            }
+        });
+        if (policyArn){
+            if (!listEntitiesForPolicy) return callback(null, results, source);
+    
+            if (listEntitiesForPolicy.err || !listEntitiesForPolicy[policyArn] || !listEntitiesForPolicy[policyArn].data) {
```
Comment (AkhtarAmir) [inline plugins/aws/iam/iamSupportPolicy.js:46]:
```suggestion
        if (policyArn){
            const listEntitiesForPolicy = helpers.addSource(cache, source,
            ['iam', 'listEntitiesForPolicy', region, policyArn]);
    
            if (!listEntitiesForPolicy || listEntitiesForPolicy.err || !listEntitiesForPolicy.data) {
```

File: plugins/aws/iam/iamSupportPolicy.js [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+                helpers.addResult(results, 3,
+                    'Unable to query for IAM entities for policy: ' + helpers.addError(listEntitiesForPolicy));
```
Comment (AkhtarAmir) [inline plugins/aws/iam/iamSupportPolicy.js:48]:
```suggestion
                    'Unable to query for IAM entities for AWSSupportAccess policy: ' + helpers.addError(listEntitiesForPolicy));
```

File: plugins/aws/iam/iamSupportPolicy.js [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+            group => attachments.push(`Policy is attached '${group.GroupName}' group`));        
+    }
+    if ('PolicyRoles' in entities) {
+        if (entities.PolicyRoles.length > 0) entities.PolicyRoles.forEach(
+            role => attachments.push(`Policy is attached '${role.RoleName}' role`));
+    }
+    if ('PolicyUsers' in entities) {
+        if (entities.PolicyGroups.length > 0) entities.PolicyUsers.forEach(
+            user => attachments.push(`Policy is attached '${user.UserName}' user`));
+    }
```
Comment (AkhtarAmir) [inline plugins/aws/iam/iamSupportPolicy.js:78]:
```suggestion
    if (entities['PolicyGroups']) {
        if (entities.PolicyGroups.length) attachments.push('group'));        
    }
    if ('PolicyRoles' in entities) {
        if (entities.PolicyRoles.length > 0) entities.PolicyRoles.forEach(
            role => attachments.push(`Policy is attached '${role.RoleName}' role`));
    }
    if ('PolicyUsers' in entities) {
        if (entities.PolicyGroups.length > 0) entities.PolicyUsers.forEach(
            user => attachments.push(`Policy is attached '${user.UserName}' user`));
    }
```

**Review by AkhtarAmir — CHANGES_REQUESTED** [review]

File: plugins/aws/iam/iamSupportPolicy.js [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+module.exports = {
+    title: 'IAM Support Policy',
+    category: 'IAM',
+    description: 'Ensures that an IAM role, group or user exists with specific permissions to access support center',
+    more_info: 'AWS provides a support center that can be used for incident notification and response, as well as technical support and customer services. An IAM Role should be present to allow authorized users to manage incidents with AWS Support',
+    link: 'https://docs.aws.amazon.com/awssupport/latest/user/accessing-support.html',
+    recommended_action: 'Ensure that an IAM role has permission to access support center.',
+    apis: ['IAM:listPolicies', 'IAM:listEntitiesForPolicy'],
+    settings: {},
```
Comment (AkhtarAmir) [inline plugins/aws/iam/iamSupportPolicy.js:11]:
```suggestion
```

File: plugins/aws/iam/iamSupportPolicy.js [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+                helpers.addResult(results, 2,
+                    'No role, user or group attached to the AWSSupportAccess policy', 'global', policyArn);
+            } else {
+                helpers.addResult(results, 0,
+                    `AWSSupportAccess Policy attached to ${attachments.join(', ')}`, 'global', policyArn);  
+            }
+        } else {
+            helpers.addResult(results, 2,
+                'No role, user or group attached to the AWSSupportAccess policy', 'global', policy.Arn);
+        }
```
Comment (AkhtarAmir) [inline plugins/aws/iam/iamSupportPolicy.js:64]:
```suggestion
        }
```

**Review by AkhtarAmir — APPROVED** [review]

**Review by giorod3 — APPROVED** [review]

**Pull request merged**